# 01 – Details

Esplorazione e data cleaning del dataset `details.csv`.

**Colonne:**
| Colonna | Descrizione |
|---|---|
| `mal_id` | ID univoco dell'anime su MyAnimeList |
| `title` | Titolo dell'anime |
| `title_japanese` | Titolo in giapponese |
| `url` | URL della pagina MAL dell'anime |
| `image_url` | URL dell'immagine di copertina |
| `type` | Tipo (TV, Movie, etc) |
| `status` | Stato di trasmissione |
| `score` | Punteggio medio |
| `scored_by` | Numero di utenti che hanno votato |
| `start_date` | Data di inizio trasmissione |
| `end_date` | Data di fine trasmissione |
| `synopsis` | Sinossi testuale |
| `rank` | Classifica globale |
| `popularity` | Popolarità |
| `members` | Numero di utenti che hanno l'anime nella loro lista |
| `favorites` | Numero di utenti che hanno messo l'anime tra i preferiti |
| `genres` | Lista di generi |
| `studios` | Lista di studi di produzione |
| `themes` | Lista di temi narrativi |
| `demographics` | Lista di target demografici |
| `source` | Materiale di origine (Manga, Novel, Original, etc) |
| `rating` | Classificazione per età |
| `episodes` | Numero di episodi |
| `season` | Stagione di uscita |
| `year` | Anno di uscita |
| `producers` | Lista di produttori |
| `explicit_genres` | Lista di generi espliciti |
| `licensors` | Lista di licenziatari |
| `streaming` | Lista di piattaforme di streaming |

## 1. Import e caricamento dati
Importiamo le librerie necessarie e carichiamo il file csv. Facciamo una esplorazione generica per capire la struttura e le caratteristiche del dataset.

In [9]:
import pandas as pd
import numpy as np
from dataset_analyzer import analyze

df_details = pd.read_csv('../datasets/details.csv')
print(f'Shape: {df_details.shape}')
print()
df_details.info()
df_details.head()

Shape: (28955, 29)

<class 'pandas.DataFrame'>
RangeIndex: 28955 entries, 0 to 28954
Data columns (total 29 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   mal_id           28955 non-null  int64  
 1   title            28955 non-null  str    
 2   title_japanese   28836 non-null  str    
 3   url              28955 non-null  str    
 4   image_url        28955 non-null  str    
 5   type             28888 non-null  str    
 6   status           28955 non-null  str    
 7   score            18882 non-null  float64
 8   scored_by        18882 non-null  float64
 9   start_date       28104 non-null  str    
 10  end_date         11267 non-null  str    
 11  synopsis         23828 non-null  str    
 12  rank             21997 non-null  float64
 13  popularity       28955 non-null  int64  
 14  members          28955 non-null  int64  
 15  favorites        28955 non-null  int64  
 16  genres           28955 non-null  str    
 17  stu

,mal_id,title,title_japanese,url,image_url,type,status,score,scored_by,start_date,...,demographics,source,rating,episodes,season,year,producers,explicit_genres,licensors,streaming
0,59356,-Socket-,-socket-,https://myanimelist.net/anime/59356/-Socket-,https://cdn.myanimelist.net/images/anime/1043/...,Movie,Finished Airing,NaN,NaN,2010-01-01T00:00:00+00:00,...,[],Original,G - All Ages,1.0,NaN,NaN,['Nagoya Zokei University'],[],[],[]
1,56036,......,......,https://myanimelist.net/anime/56036/-,https://cdn.myanimelist.net/images/anime/1057/...,Music,Finished Airing,6.53,503.0,2023-06-11T00:00:00+00:00,...,[],Original,PG-13 - Teens 13 or older,1.0,NaN,NaN,[],[],[],[]
2,2928,.hack//G.U. Returner,.HACK//G.U. RETURNER,https://myanimelist.net/anime/2928/hack__GU_Re...,https://cdn.myanimelist.net/images/anime/1798/...,OVA,Finished Airing,6.65,9745.0,2007-01-18T00:00:00+00:00,...,[],Game,PG-13 - Teens 13 or older,1.0,NaN,NaN,"['Bandai Visual', 'CyberConnect2']",[],[],[]
3,3269,.hack//G.U. Trilogy,.hack//G.U. Trilogy,https://myanimelist.net/anime/3269/hack__GU_Tr...,https://cdn.myanimelist.net/images/anime/1566/...,Movie,Finished Airing,7.06,15373.0,2007-12-22T00:00:00+00:00,...,[],Game,PG-13 - Teens 13 or older,1.0,NaN,NaN,['Bandai Visual'],[],"['Funimation', 'Bandai Entertainment']",[]
4,4469,.hack//G.U. Trilogy: Parody Mode,.hack//G.U. Trilogy,https://myanimelist.net/anime/4469/hack__GU_Tr...,https://cdn.myanimelist.net/images/anime/10/86...,Special,Finished Airing,6.35,4317.0,2008-03-25T00:00:00+00:00,...,[],Game,PG-13 - Teens 13 or older,1.0,NaN,NaN,['Bandai Visual'],[],[],[]


Il dataset contiene **28.955 righe** e **29 colonne**. Alcune colonne presentano valori nulli. Alcuni dtype richiedono conversione: le date sono in formato stringa, e alcune colonne intere (`scored_by`, `rank`, `episodes`, `year`) sono memorizzate come `float64`. Quiesti veranno analizzati e gestiti singolarmente nella sezione 2.

## 1.1 Rimozione duplicati esatti

Prima dell'analisi per colonna, rimuoviamo le righe con valori identici in **tutte** le colonne, mantenendo solo la prima occorrenza.

In [10]:
n_originale = len(df_details)

mask_dup = df_details.duplicated(keep=False)
n_righe_coinvolte = mask_dup.sum()
n_gruppi = df_details[mask_dup].duplicated(keep='first').sum()
n_tenute = n_righe_coinvolte - n_gruppi

print(f'Righe totali coinvolte in duplicazioni : {n_righe_coinvolte:,}')
print(f'  → prime occorrenze mantenute         : {n_tenute:,}')
print(f'  → occorrenze extra rimosse           : {n_gruppi:,}')
print()

df_details.drop_duplicates(keep='first', inplace=True)
print(f'Righe prima della rimozione : {n_originale:,}')
print(f'Righe dopo la rimozione     : {len(df_details):,}')

Righe totali coinvolte in duplicazioni : 0
  → prime occorrenze mantenute         : 0
  → occorrenze extra rimosse           : 0

Righe prima della rimozione : 28,955
Righe dopo la rimozione     : 28,955


## 2. Analisi colonna per colonna

### 2.1 `mal_id`

È la **chiave primaria** del dataset. Deve essere intera, non nulla e univoca. Utilizziamo `analyze` per verificare queste proprietà.

In [11]:
analyze(df_details['mal_id'])


  Nome serie                     mal_id
  dtype                          int64
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           NUMERICO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,955  (100.00%)
  Null / NaN                     0  (0.00%)
  Zeri                           0  (0.00%)
  Positivi                       28,955   (100.00%)
  Negativi                       0   (0.00%)
  Valori unici                   28,955  (100.00%)

Statistiche descrittive
────────────────────────────────────────────────────────────────────────────────

  Min                            1
  Max                            62,590
  Range                          62,589
  Media                          33,977.52
  Mediana                        37,969.00
  Moda/e                         1, 5, 6
  Dev. standard    

**Osservazioni:**
- **Nessun valore nullo**: tutti i 28.955 record hanno un ID univoco.
- **Nessun duplicato**: ogni riga corrisponde a un anime distinto.
- **Dtype corretto**: `int64`, nessuna conversione necessaria.

**Nessuna pulizia necessaria.**

### 2.2 `title`

Colonna testuale con il titolo dell'anime. Verifichiamo la presenza di valori nulli e duplicati.

In [12]:
df_details['title'] = df_details['title'].str.strip()
analyze(df_details['title'])


  Nome serie                     title
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,955  (100.00%)
  Null / NaN                     0  (0.00%)
    di cui stringhe vuote        0  (convertite a NULL)
  Valori unici                   28,953  (99.99%)
  Valori duplicati               2  (0.01%)

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  1  caratteri
  Lunghezza max                  186  caratteri
  Lunghezza media                24.70  caratteri
  Lunghezza mediana              21.0000  caratteri
  Dev. standard lunghezza        15.07
  IQR lunghezza                  18.0000

Valori di lunghez

**Osservazioni:**
- **Nessun valore nullo**: tutti i 28.955 record hanno un titolo.
- **Quasi univoco**: 28.953 valori distinti su 28.955. Ci solo 2 titoli duplicati che verifichiamo successivamente.
- **Lunghezze variabili**: da 1 carattere a 186 caratteri. Stiampiamo i 10 più lunghi e 10 più corti e facciamo delle verifiche per approfondire.
- **14 titoli solo numerici**: Indaghiamo in seguito.
- **1 titolo contiene un URL**: Ingaghiamo in seguito.
- **43.4% contiene caratteri speciali**: normale, i titoli includono punteggiatura, simboli e trattini.

In [13]:
# Titoli duplicati
dup_mask = df_details.duplicated(subset=['title'], keep=False)
print(f'Titoli duplicati (righe coinvolte): {dup_mask.sum()}')
print()
print(df_details[dup_mask][['mal_id', 'title', 'type', 'status', 'year']].to_string(index=False))

# Titoli solo numerici
print()
print('─' * 60)
num_mask = df_details['title'].str.fullmatch(r'\d+')
print(f'Titoli solo numerici: {num_mask.sum()}')
print()
print(df_details[num_mask][['mal_id', 'title', 'type', 'year']].to_string(index=False))

# Titolo che contiene un URL
print()
print('─' * 60)
url_mask = df_details['title'].str.contains(r'https?://', regex=True, na=False)
print(f'Titoli con URL: {url_mask.sum()}')
print()
print(df_details[url_mask][['mal_id', 'title', 'type', 'year']].to_string(index=False))

# Lunghezze estreme
print()
print('─' * 60)
lunghezze = df_details['title'].str.len()
top10_lunghi = df_details.loc[lunghezze.nlargest(10).index][['mal_id', 'title', 'type', 'year']]
top10_corti  = df_details.loc[lunghezze.nsmallest(10).index][['mal_id', 'title', 'type', 'year']]
print('10 titoli più lunghi:')
print(top10_lunghi.to_string(index=False))
print()
print('10 titoli più corti:')
print(top10_corti.to_string(index=False))

# Titoli con caratteri speciali
print()
print('─' * 60)
spec_mask = df_details['title'].str.contains(r'[^a-zA-Z0-9\s]', regex=True, na=False)
print(f'Titoli con caratteri speciali: {spec_mask.sum()}')
print()
print('Primi 10:')
print(df_details[spec_mask][['mal_id', 'title', 'type', 'year']].head(10).to_string(index=False))

Titoli duplicati (righe coinvolte): 4

 mal_id                                    title  type          status  year
  59994                                Moonlight Music Finished Airing   NaN
  60524                                Moonlight Music Finished Airing   NaN
  53367 Shen Lan Qi Yu Wushuang Zhu: Tianmo Pian   ONA Finished Airing   NaN
  55740 Shen Lan Qi Yu Wushuang Zhu: Tianmo Pian   ONA Finished Airing   NaN

────────────────────────────────────────────────────────────
Titoli solo numerici: 14

 mal_id  title  type   year
  29978    001   ONA    NaN
  60145   1135 Music    NaN
  37065    123 Music    NaN
  55204   1985 Music    NaN
  30234   1989 Music    NaN
  38756      2 Music    NaN
  54066   2005 Music    NaN
  32813   2010 Music    NaN
  53789     32 Music    NaN
  53771   3341 Music    NaN
  60486    520 Movie    NaN
  13971 663114 Movie    NaN
  41457     86    TV 2021.0
  36089      9 Music    NaN

────────────────────────────────────────────────────────────
Titoli

**Risultati dell'indagine:**

- **Titoli duplicati (2 coppie, 4 righe):** `Moonlight` e `Shen Lan Qi Yu Wushuang Zhu: Tianmo Pian` compaiono ciascuno due volte, ma con `mal_id` distinti. Sono anime separati su MAL che condividono il titolo. **Nessuna rimozione necessaria.**

- **Titoli solo numerici (14 righe):** tutti legittimi. La maggior parte sono video musicali (`Music`) o cortometraggi in cui titoli brevi o numerici sono comuni. **Nessuna azione necessaria.**

- **Titolo contenente un URL (1 riga):** `Himitsukessha Taka no Tsume The Movie 3: http://takanotsume.jp wa Eien ni` (`mal_id` 8357). L'URL è parte integrante del titolo ufficiale del film, caratteristica della serie *Taka no Tsume* nota per titoli volutamente assurdi. **Nessuna azione necessaria.**

- **Lunghezze estreme:** i titoli molto corti (1–2 caratteri) sono anime reali con titoli brevi; i titoli molto lunghi (fino a 186 caratteri) sono tipici degli produzioni moderni. Nessun valore anomalo.

- 43.4% dei titoli contiene caratteri speciali (12.572): percentuale alta ma attesa. Dai primi 10 stampati si nota che si tratta di punteggiatura e simboli che fanno   parte naturale titoli anime.

**Conclusione:** la colonna `title` non richiede alcuna pulizia. Tutti i casi apparentemente anomali sono valori legittimi.

### 2.3 `title_japanese`

Colonna testuale con il titolo in giapponese. I null sono ammessi: anime di produzione non giapponese o senza titolo kanji registrato su MAL non avranno questo campo.

In [14]:
df_details['title_japanese'] = df_details['title_japanese'].str.strip()
analyze(df_details['title_japanese'])


  Nome serie                     title_japanese
  dtype                          str
  Dimensione                     28,955
  Range indice                   0 … 28954
  Tipo                           STRINGA / OGGETTO

Conteggi di base
────────────────────────────────────────────────────────────────────────────────

  Righe totali                   28,955
  Valori non nulli               28,836  (99.59%)
  Null / NaN                     119  (0.41%)
    di cui stringhe vuote        0  (convertite a NULL)
  Valori unici                   27,772  (95.91%)
  Valori duplicati               1,064  (3.67%)

Analisi lunghezza stringhe
────────────────────────────────────────────────────────────────────────────────

  Lunghezza min                  1  caratteri
  Lunghezza max                  85  caratteri
  Lunghezza media                12.24  caratteri
  Lunghezza mediana              10.0000  caratteri
  Dev. standard lunghezza        8.69
  IQR lunghezza                  10.0000

Valor

**Osservazioni:**
- **119 valori nulli** (~0.4%): strutturali, anime di produzione non giapponese o senza titolo kanji registrato su MAL.
- Facciamo un controllo sui titoli duplicati, con lunghezze estreme, contenenti `@` e solo numerici.

In [15]:
cols = ['mal_id', 'title', 'title_japanese', 'type', 'year']

# Duplicati
dup_mask_tj = df_details['title_japanese'].duplicated(keep=False) & df_details['title_japanese'].notna()
print(f'Titoli giapponesi duplicati (righe coinvolte): {dup_mask_tj.sum()}')
print()
print('Primi 10:')
print(df_details[dup_mask_tj][cols].head(10).to_string(index=False))

# Lunghezze estreme
print()
print('─' * 60)
lunghezze_tj = df_details['title_japanese'].str.len()
print('10 titoli giapponesi più lunghi:')
print(df_details.loc[lunghezze_tj.nlargest(10).index, cols].to_string(index=False))
print()
print('10 titoli giapponesi più corti:')
print(df_details.loc[lunghezze_tj.nsmallest(10).index, cols].to_string(index=False))

# Titoli con @
print()
print('─' * 60)
at_mask_tj = df_details['title_japanese'].str.contains('@', regex=False, na=False)
print(f'Titoli giapponesi con @: {at_mask_tj.sum()}')
print()
print(df_details[at_mask_tj][cols].to_string(index=False))

# Solo numerici
print()
print('─' * 60)
num_mask_tj = df_details['title_japanese'].str.fullmatch(r'\d+', na=False)
print(f'Titoli giapponesi solo numerici: {num_mask_tj.sum()}')
print()
print(df_details[num_mask_tj][cols].to_string(index=False))

Titoli giapponesi duplicati (righe coinvolte): 2003

Primi 10:
 mal_id                                                 title      title_japanese       type   year
   3269                                   .hack//G.U. Trilogy .hack//G.U. Trilogy      Movie    NaN
   4469                      .hack//G.U. Trilogy: Parody Mode .hack//G.U. Trilogy    Special    NaN
  41380            100-man no Inochi no Ue ni Ore wa Tatteiru    100万の命の上に俺は立っている         TV 2020.0
  44881 100-man no Inochi no Ue ni Ore wa Tatteiru 2nd Season    100万の命の上に俺は立っている         TV 2021.0
  48976      100-man no Inochi no Ue ni Ore wa Tatteiru Recap    100万の命の上に俺は立っている TV Special    NaN
  32601                   12-sai.: Chicchana Mune no Tokimeki  12歳。～ちっちゃなムネのトキメキ～         TV 2016.0
  33419        12-sai.: Chicchana Mune no Tokimeki 2nd Season  12歳。～ちっちゃなムネのトキメキ～         TV 2016.0
   9624                                30-sai no Hoken Taiiku            30歳の保健体育         TV 2011.0
  12131                       30-sai 

- I titoli giaponesi contenenti `@` corrispondono con titoli in inglese con `@` nella stessa posizione. Il carattere fa parte del titolo. Nessuna pulizia necessaria.
- I titoli giaponesi solo numerici corrispondono a titoli in inglese anche loro solo numerici. Nessuna pulizia necessaria.
- I titoli più corti come osservato dai primi 10 stampati mostrano una corrispondenza con i titoli in inglese e appartengono principalmente a brani musicali o film. Anche questo è atteso.
- I titoli più lunghi seguono la stessa logica dei titoli più corti con l'eccezione di un titolo che è uguale a quello in inglese. Facciamo un controllo per vedere se è un caso isolato o se accade più volte.

In [16]:
# title == title_japanese
same_mask = df_details['title'] == df_details['title_japanese']
print(f'Titoli identici in title e title_japanese: {same_mask.sum()} ({same_mask.mean()*100:.1f}%)')
print()
print('Primi 10:')
print(df_details[same_mask][['mal_id', 'title', 'title_japanese', 'type', 'year']].head(10).to_string(index=False))

Titoli identici in title e title_japanese: 833 (2.9%)

Primi 10:
 mal_id                 title        title_japanese    type   year
  56036                ......                ......   Music    NaN
   3269   .hack//G.U. Trilogy   .hack//G.U. Trilogy   Movie    NaN
   1143     .hack//Intermezzo     .hack//Intermezzo Special    NaN
   9332        .hack//Quantum        .hack//Quantum     OVA    NaN
    873          .hack//Roots          .hack//Roots      TV 2006.0
   1487         .hack//Unison         .hack//Unison Special    NaN
  29978                   001                   001     ONA    NaN
  28761                 00:08                 00:08   Movie    NaN
   7374                  100%                  100%     OVA    NaN
  39208 100% Renewable Energy 100% Renewable Energy Special    NaN


Notiamo che ci sono 833 righe con nome in inglese uguale a quello in giaponese. Potrebbe trattarsi di opere senza nome giaponese. Abbiamo deciso di tenerlo invariato.

**Nessuna pulizia necessaria**